Jenkinsfile

```diff
@@ -0,0 +1,141 @@
+    wrap([$class: 'TimestamperBuildWrapper']) {
+
+        // First stage is actually checking out the source. Since we're using Multibranch
+        // currently, we can use "checkout scm".
+        stage "Checkout source"
+
+        checkout scm
+
+        // Now run the actual build.
+        stage "Build and test"
+
+        // We're wrapping this in a timeout - if it takes more than 180 minutes, kill it.
+        timeout(time: 180, unit: 'MINUTES') {
+            // See below for what this method does - we're passing an arbitrary environment
+            // variable to it so that JAVA_OPTS and MAVEN_OPTS are set correctly.
+            withMavenEnv(["JAVA_OPTS=-Xmx1536m -Xms512m -XX:MaxPermSize=1024m",
+                          "MAVEN_OPTS=-Xmx1536m -Xms512m -XX:MaxPermSize=1024m"]) {
+                // Actually run Maven!
+                // The -Dmaven.repo.local=${pwd()}/.repository means that Maven will create a
+                // .repository directory at the root of the build (which it gets from the
+                // pwd() Workflow call) and use that for the local Maven repository.
+                sh "mvn -Pdebug -U clean install ${runTests ? '-Dmaven.test.failure.ignore=true -Dconcurrency=1' : '-DskipTests'} -V -B -Dmaven.repo.local=${pwd()}/.repository"
+            }
+        }
+
+        // Once we've built, archive the artifacts and the test results.
+        stage "Archive artifacts and test results"
+
+        archive includes: "**/target/*.jar, **/target/*.war, **/target/*.hpi"
+        if (runTests) {
+            step([$class: 'JUnitResultArchiver', healthScaleFactor: 20.0, testResults: '**/target/surefire-reports/*.xml'])
+        }
+
+        // And stash the jenkins.war for the next step
+        stash name: "jenkins.war", includes: "war/target/jenkins.war"
+    }
+}
+
+// Run the packaging build on a node with the "pkg" label.
+node('docker') {
+    // Add timestamps to logging output.
+    wrap([$class: 'TimestamperBuildWrapper']) {
+
+        // First stage here is getting prepped for packaging.
+        stage "packaging - docker prep"
+
+        // Docker environment to build packagings
+        dir('packaging-docker') {
+            git branch: packagingBranch, url: 'https://github.com/jenkinsci/packaging.git'
+            sh 'docker build -t jenkins-packaging-builder:0.1 docker'
+        }
+
+        stage "packaging - actually packaging"
+        // Working packaging code, separate branch with fixes
+        dir('packaging') {
+            git branch: packagingBranch, url: 'https://github.com/jenkinsci/packaging.git'
+            // Grab the war file from the stash - it goes to war/target/jenkins.war
+            unstash "jenkins.war"
+            sh "cp war/target/jenkins.war ."
+
+            sh 'docker run --rm -v "`pwd`":/tmp/packaging -w /tmp/packaging jenkins-packaging-builder:0.1 make clean deb rpm suse BRAND=./branding/jenkins.mk BUILDENV=./env/test.mk CREDENTIAL=./credentials/test.mk WAR=jenkins.war'
```

Comment:
I'm pretty sure this can be obviated with the [docker workflow plugin](http://documentation.cloudbees.com/docs/cje-user-guide/docker-workflow-sect-inside.html), e.g 

```groovy
docker.image('jenkins-packaging-builder:0.1').inside { c ->
  sh 'mvn'
}
```